expr/functions/weightedAverage/function.go

```diff
@@ -0,0 +1,163 @@
+package weightedAverage
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/bookingcom/carbonapi/expr/functions/sum"
+	"github.com/bookingcom/carbonapi/expr/helper"
+	"github.com/bookingcom/carbonapi/expr/interfaces"
+	"github.com/bookingcom/carbonapi/expr/types"
+	"github.com/bookingcom/carbonapi/pkg/parser"
+)
+
+type weightedAverage struct {
+	interfaces.FunctionBase
+}
+
+func GetOrder() interfaces.Order {
+	return interfaces.Any
+}
+
+func New(configFile string) []interfaces.FunctionMetadata {
+	res := make([]interfaces.FunctionMetadata, 0)
+	f := &weightedAverage{}
+	functions := []string{"weightedAverage"}
+	for _, n := range functions {
+		res = append(res, interfaces.FunctionMetadata{Name: n, F: f})
+	}
+	return res
+}
+
+// exclude(seriesList, pattern)
+func (f *weightedAverage) Do(ctx context.Context, e parser.Expr, from, until int32, values map[parser.MetricRequest][]*types.MetricData, getTargetData interfaces.GetTargetData) ([]*types.MetricData, error) {
+	avgArg, err := helper.GetSeriesArg(ctx, e.Args()[0], from, until, values, getTargetData)
+	if err != nil {
+		return nil, err
+	}
+	weightArg, err := helper.GetSeriesArg(ctx, e.Args()[1], from, until, values, getTargetData)
+	if err != nil {
+		return nil, err
+	}
+	nodes, err := e.GetIntArgs(2)
+	if err != nil {
+		return nil, err
+	}
+
+	avgGrouped, avgNodeKeys, err := helper.GroupByNodes(nodes, avgArg)
+	if err != nil {
+		return nil, err
+	}
+	weightGrouped, weightNodeKeys, err := helper.GroupByNodes(nodes, weightArg)
+	if err != nil {
+		return nil, err
+	}
+
+	var keys []string
+	for avgKey := range avgNodeKeys {
+		if _, found := weightNodeKeys[avgKey]; found {
+			keys = append(keys, avgKey)
+		}
+	}
+
+	var productMetrics []*types.MetricData
+	for _, key := range keys {
+		if len(avgGrouped[key]) == 0 || len(weightGrouped[key]) == 0 {
+			continue
+		}
+		weight := weightGrouped[key][len(weightGrouped[key])-1]
+		avg := avgGrouped[key][len(avgGrouped[key])-1]
+		productMetricName := fmt.Sprintf("product(%s, %s)", weight.Name, avg.Name)
+		productMetric := helper.CombineSeries(avg, weight, productMetricName, productOperator)
```

Comment:
is there any specific reason that you are using `helper.CombineSeries` rather than `helper.AggregateSeries` here?

Reply:
well no. Since it was only two timeseries for the operation, I used `helper.CombineSeries`. is there a concern in using it?

Reply:
I was reading both functions and it seems `helper.CombineSeries` and `helper.AggregateSeries` are doing similar things and `helper.CombineSeries` doesn't exist in `go-graphite/carbonapi`. Probably a refactoring target.